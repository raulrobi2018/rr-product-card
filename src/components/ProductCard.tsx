import React from 'react';

import styles from '../styles/styles.module.css';

import { useProduct } from '../hooks/useProduct';
import { createContext } from 'react';
import { ProductCardHandlers } from '../interfaces/interfaces';
import {
  ProductContextProps,
  Product,
  onChangeArgsProps,
  InitialValuesProps,
} from '../interfaces/interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;
export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  onChange?: (args: onChangeArgsProps) => void;
  value?: number;
  initialValues?: InitialValuesProps;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider value={{ counter, increaseBy, maxCount, product }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {//React no permite renderizar funciones, pero podemos ejecutarla de la manera
        //en que la estamos haciendo aqu√≠ ya que lo que hace es devolver un JSX
        children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
