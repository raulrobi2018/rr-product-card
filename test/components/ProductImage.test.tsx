import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('should display the component correctly without image', () => {
    const wrapper = renderer.create(<ProductImage img="http://image.jpg" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
    console.log(wrapper.toJSON());
  });

  test('should display the product with image', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
