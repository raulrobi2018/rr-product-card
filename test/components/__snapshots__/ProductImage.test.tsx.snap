// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should display the component correctly without image 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "center",
    }
  }
>
  <img
    alt="Product"
    className="productImg undefined"
    src="http://image.jpg"
  />
</div>
`;

exports[`ProductImage should display the product with image 1`] = `
<div
  className="productCard undefined"
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <img
      alt="Product"
      className="productImg undefined"
      src="./coffee-mug.png"
    />
  </div>
</div>
`;
